File: src/main/scala/com/learning/scala/classobj/CashFlow.scala
package com.learning.scala.classobj

/* Scala: первичные и вспомогательные конструкторы */

/* класс имеющий первичный конструктор, который определяет параметры */
class CashFlow(amt: Double, curr: String) { /* первичный конструктор с двумя аргументами */

  /* класс может иметь дополнительные вспомогательные конструкторы, которые действуют как поставщиков-параметров по умолчанию для первичного конструктора */
  def this(amt: Double) = this(amt, "INR")
  def this(curr: String) = this(0, curr)

  /* определять методы без включения в скобки или список параметров (даже пустой): */
  def amount = amt
  def currency = curr
}

object CashFlow {
  def main(args: Array[String]) {
    /* пример создание объекта с разными передаваемыми параметрами: */
//    val cf = new CashFlow(500.0)
//    val cf = new CashFlow("Aaa")
    val cf = new CashFlow(1000.0, "Bbb")

    println( "Amount = " + cf.amount )
    println( "Currency: " + cf.currency )
  }
}

File: src/main/java/com/learning/java/classobj/CashFlow.java
package com.learning.java.classobj;

/* Java: перегрузка (overloaded) конструктора */

public class CashFlow {
	private final Double amt;
	private final String curr;

	public CashFlow(Double amt, String curr) {
		this.amt = amt;
		this.curr = curr;
	}
	
	public CashFlow(Double amt) {
		this(amt, "INR");
	}
	
	public CashFlow(String curr) {
		this(0.0, curr);
	}

	public Double getAmount() {
		return amt;
	}
	
	public String getCurrency() {
		return curr;
	}

	public static void main(String[] args) {
		CashFlow cf = new CashFlow(500.0);

		System.out.println( "Amount = " + cf.getAmount() );
		System.out.println( "Currency: " + cf.getCurrency() );
	}
}
